File: py/navis_fastcore/elastix.py
"""Applying Elastix transformations to points."""

import os

import numpy as np

from . import _fastcore
from ._points import _prep_points

__all__ = ["ElastixTransform", "load_elastix_transform"]


class ElastixTransform:
    """An Elastix transform - or a chain of them - ready to apply to points.

    The transform is parsed once, on construction, and can then be applied any number of
    times. A ``TransformParameters`` file is *already* a chain: its
    ``InitialTransformParametersFileName`` is followed recursively (resolved relative to
    that file's own directory), so a four-deep affine -> affine -> B-spline -> B-spline
    stack loads from its outermost file alone.

    Unlike `navis`, this does not shell out to the ``transformix`` binary: **Elastix does
    not need to be installed**, and there is no subprocess, no temporary directory and no
    ``copy_files`` dance. Results match ``transformix`` to ~5e-7 - its own print precision.

    It also gives you an **inverse**, which Elastix has no way to compute. See
    :meth:`~navis_fastcore.ElastixTransform.xform_inv`.

    Parameters
    ----------
    path :      str | pathlib.Path | list thereof
                Path to a ``TransformParameters.*.txt`` file. Pass a list to build a chain,
                applied in order: ``points -> path[0] -> path[1] -> ... -> output``.
    invert :    bool | list of bool
                Traverse a transform backwards. A single bool applies to every entry; pass
                a list to set them per transform. Useful when routing through a bridging
                graph, where an edge may be traversed in either direction.

    Attributes
    ----------
    affine :        (4, 4) np.ndarray
                    The matrix of the first linear step of the first transform, if it has
                    one; ``None`` otherwise.
    kinds :         list of list of str
                    The resolved step kinds of each transform, initial first - e.g.
                    ``[["linear", "bspline"]]``.
    grid_size :     (k, 3) np.ndarray
                    Control-point grid size of every B-spline in the chain.
    grid_spacing :  (k, 3) np.ndarray
                    Control-point spacing of every B-spline in the chain.
    grid_origin :   (k, 3) np.ndarray
                    Control-point grid origin of every B-spline in the chain.

    Examples
    --------
    >>> import navis_fastcore as fastcore
    >>> import numpy as np
    >>> xf = fastcore.ElastixTransform("TransformParameters.FixedFANC.txt")  # doctest: +SKIP
    >>> pts = np.array([[50.0, 50.0, 50.0], [100.0, 200.0, 40.0]])
    >>> out = xf.xform(pts)                                                  # doctest: +SKIP
    >>> back = xf.xform_inv(out)                                             # doctest: +SKIP
    >>> np.allclose(xf.xform(back), out, atol=1e-3)                          # doctest: +SKIP
    True

    """

    def __init__(self, path, invert=False):
        if isinstance(path, (str, os.PathLike)):
            paths = [os.fspath(path)]
        else:
            paths = [os.fspath(p) for p in path]
        if not paths:
            raise ValueError("`path` must name at least one transform")

        if isinstance(invert, bool):
            inv = [invert] * len(paths)
        else:
            inv = [bool(i) for i in invert]
            if len(inv) != len(paths):
                raise ValueError(
                    f"`invert` must have one flag per transform: expected "
                    f"{len(paths)}, got {len(inv)}"
                )

        self._paths = paths
        self._invert = inv
        self._xf = _fastcore.ElastixTransform(paths, inv)

    def xform(self, points, out_of_bounds="identity", n_cores=None, progress=False):
        """Transform points forward.

        Parameters
        ----------
        points :            (N, 3) array | (3,) array
                            Coordinates to transform.
        out_of_bounds :     "identity" | "nan"
                            What to do with points that fall outside a B-spline's
                            control-point grid. ``"identity"`` (the default) returns them
                            **unchanged**, which is exactly what ``transformix`` does.
                            ``"nan"`` returns ``NaN`` instead.

                            The default is silent by nature: a neuron straddling the grid
                            edge comes back partly transformed and looks perfectly fine.
                            Use ``"nan"`` when you would rather see the boundary than trust
                            it.
        n_cores :           int, optional
                            Cap the thread pool. ``None`` uses all cores.
        progress :          bool
                            Show a progress bar.

        Returns
        -------
        (N, 3) np.ndarray
                    Transformed coordinates. A ``(3,)`` input gives a ``(3,)`` output.

        """
        pts, was_1d = _prep_points(points)
        out = self._xf.xform(
            pts,
            str(out_of_bounds),
            None if n_cores is None else int(n_cores),
            bool(progress),
        )
        return out[0] if was_1d else out

    def xform_inv(
        self,
        points,
        out_of_bounds="identity",
        initial_guess=None,
        max_iter=50,
        seed_iter=8,
        tolerance=1e-9,
        accuracy=1e-3,
        lattice_points=16_000,
        n_cores=None,
        progress=False,
    ):
        """Transform points backwards - something Elastix itself cannot do.

        Linear steps are inverted exactly. Each B-spline warp has no closed-form inverse
        and is solved per point by damped Gauss-Newton against the analytic Jacobian.

        What is guaranteed is **forward-consistency**: ``xform(xform_inv(y)) == y``, to
        within ``accuracy``. What is *not* guaranteed is that ``xform_inv(xform(p)) == p``,
        because a B-spline warp need not be injective - a strongly folded registration maps
        several points to the same place, and no inverse can recover which one you meant.
        Points with no preimage at all come back as ``NaN``.

        Parameters
        ----------
        points :            (N, 3) array | (3,) array
                            Coordinates to transform.
        out_of_bounds :     "identity" | "nan"
                            See :meth:`~navis_fastcore.ElastixTransform.xform`.
        initial_guess :     (N, 3) array, optional
                            Starting points for the solver. Rarely needed: it seeds itself
                            with a fixed-point iteration, which is what makes it converge
                            even where the deformation is large.
        max_iter :          int
                            Solver budget per point.
        seed_iter :         int
                            Rounds of the fixed-point pre-seed. Zero starts the solver at
                            the target, which fails wherever the deformation is large.
        tolerance :         float
                            Step-size convergence threshold.
        accuracy :          float
                            Accept a solution only if its residual is within this of the
                            target, in world units. Otherwise the row is ``NaN``.
        lattice_points :    int
                            Size of the global seed lattice - the last-resort start for the
                            few points the cheap seeds fail on. Built once per call, and only
                            consulted by points that have already failed, so it costs almost
                            nothing on a well-behaved registration. Set to 0 to disable.
        n_cores :           int, optional
                            Cap the thread pool. ``None`` uses all cores.
        progress :          bool
                            Show a progress bar.

        Returns
        -------
        (N, 3) np.ndarray
                    Coordinates in the source space. Rows with no preimage are ``NaN``.

        """
        pts, was_1d = _prep_points(points)
        guess = None
        if initial_guess is not None:
            guess, _ = _prep_points(initial_guess, name="initial_guess")
            if guess.shape != pts.shape:
                raise ValueError(
                    "`initial_guess` must have one point per input point: expected "
                    f"{pts.shape}, got {guess.shape}"
                )
        out = self._xf.xform_inv(
            pts,
            str(out_of_bounds),
            guess,
            int(max_iter),
            int(seed_iter),
            float(tolerance),
            float(accuracy),
            int(lattice_points),
            None if n_cores is None else int(n_cores),
            bool(progress),
        )
        return out[0] if was_1d else out

    @property
    def affine(self):
        """The (4, 4) matrix of the first linear step, or `None`."""
        return self._xf.affine

    @property
    def kinds(self):
        """The step kinds of each transform in the chain, initial first."""
        return self._xf.kinds

    @property
    def invertible(self):
        """Whether `xform_inv` can run. False only for a chain carrying an ``Add`` step."""
        return self._xf.invertible

    @property
    def grid_size(self):
        """Control-point grid size of every B-spline in the chain, `(k, 3)`."""
        return self._xf.grid_size

    @property
    def grid_spacing(self):
        """Control-point spacing of every B-spline in the chain, `(k, 3)`."""
        return self._xf.grid_spacing

    @property
    def grid_origin(self):
        """Control-point grid origin of every B-spline in the chain, `(k, 3)`."""
        return self._xf.grid_origin

    @property
    def paths(self):
        """The files this transform was loaded from."""
        return list(self._paths)

    def __len__(self):
        return self._xf.n_transforms

    def __reduce__(self):
        # Re-load from disk in the child rather than pickling the coefficients through every
        # multiprocessing/joblib fan-out. BANC's `BANC_to_template.txt` is 56 MB.
        return (ElastixTransform, (self._paths, self._invert))

    def __repr__(self):
        return self._xf.__repr__()


def load_elastix_transform(path, invert=False):
    """Load one or more Elastix transforms.

    A convenience wrapper around :class:`~navis_fastcore.ElastixTransform`.

    Parameters
    ----------
    path :      str | pathlib.Path | list thereof
                Path to a ``TransformParameters.*.txt`` file, or several to chain.
    invert :    bool | list of bool
                Traverse a transform backwards.

    Returns
    -------
    ElastixTransform

    """
    return ElastixTransform(path, invert=invert)

File: R/nat.fastcore/R/elastix.R
#' Read an Elastix transform
#'
#' Reads an Elastix `TransformParameters` file, ready to apply to points. **Elastix itself does
#' not need to be installed** -- there is no call out to `transformix`, no subprocess and no
#' temporary directory.
#'
#' A `TransformParameters` file is already a chain: its `InitialTransformParametersFileName` is
#' followed recursively, resolved relative to *that file's own directory*. So a four-deep
#' affine -> affine -> B-spline -> B-spline stack loads from its outermost file alone.
#'
#' The transform is parsed once and can then be applied any number of times, which is what you
#' want when transforming every neuron in a dataset.
#'
#' @param path Path to a `TransformParameters.*.txt` file. Pass several to build a chain, applied
#'   in order: `xyz -> path[1] -> path[2] -> ... -> output`.
#'
#' Direction is **not** fixed here. It is chosen per call -- see the `invert` argument of
#' [elastix_xform()], and [elastix_xform_inv()] -- so one object serves every direction and the
#' file is parsed only once. That is worth caring about: real warps run to tens of megabytes.
#'
#' @return An object of class `elastix_transform`.
#'
#' @seealso [elastix_xform()], [elastix_xform_inv()]
#'
#' @examples
#' \dontrun{
#' xf <- elastix_read("TransformParameters.FixedFANC.txt")
#' xyzmatrix(n) <- elastix_xform(xf, xyzmatrix(n))
#' }
#'
#' @export
elastix_read <- function(path) {
  path <- as.character(path)
  if (!length(path)) stop("`path` must name at least one transform")

  # Validate here rather than in Rust. extendr cannot raise an R condition from a constructor --
  # a panic there loses its message and R only sees "User function panicked" -- so the common
  # failure (a missing file) has to be caught on this side to produce a message worth reading.
  for (p in path) {
    if (!file.exists(p) || dir.exists(p)) {
      stop(sprintf("no Elastix transform file at '%s'", p))
    }
  }

  ptr <- ElastixTransformPtr$load(path)
  structure(list(ptr = ptr, path = path), class = "elastix_transform")
}

#' @export
print.elastix_transform <- function(x, ...) {
  kinds <- x$ptr$kinds()
  cat(sprintf(
    "<elastix_transform> %d transform%s: %s\n",
    length(kinds), if (length(kinds) == 1L) "" else "s", paste(kinds, collapse = " -> ")
  ))
  for (i in seq_along(x$path)) {
    cat(sprintf("  %s\n", x$path[i]))
  }
  invisible(x)
}

.elastix_ptr <- function(xf) {
  if (!inherits(xf, "elastix_transform")) {
    stop("`xf` must be an <elastix_transform>, as returned by `elastix_read()`")
  }
  xf$ptr
}

#' Apply an Elastix transform to points
#'
#' @param xf An `elastix_transform`, from [elastix_read()].
#' @param xyz An `(N, 3)` matrix of coordinates. A data frame is also accepted -- if it has
#'   x/y/z columns (in any case) those are used, whatever else it carries, otherwise the
#'   first three columns -- as is a bare length-3 vector.
#' @param out_of_bounds What to do with points outside a B-spline's control-point grid.
#'   `"identity"` (the default) returns them **unchanged**, which is exactly what `transformix`
#'   does. `"nan"` returns `NaN` instead.
#'
#'   The default is silent by nature: a neuron straddling the grid edge comes back partly
#'   transformed and looks perfectly fine. Use `"nan"` when you would rather see the boundary
#'   than trust it.
#' @param invert Logical, length 1 or one flag per transform. `TRUE` traverses that transform
#'   backwards -- what you need when routing through a bridging graph, where an edge may be walked
#'   in either direction.
#'
#'   This is **not** the same as [elastix_xform_inv()], which inverts the whole composition
#'   (reversing the order *and* flipping every hop). For a single transform the two agree; for a
#'   chain they do not, and only `invert` can express a mixed-direction traversal.
#' @param n_cores Cap the thread pool. `NULL` uses all cores.
#' @param progress Show a progress bar.
#'
#' @return An `(N, 3)` matrix of transformed coordinates.
#'
#' @seealso [elastix_xform_inv()]
#'
#' @examples
#' \dontrun{
#' xf <- elastix_read("TransformParameters.FixedFANC.txt")
#' elastix_xform(xf, cbind(50, 50, 50))
#' }
#'
#' @export
elastix_xform <- function(xf, xyz, out_of_bounds = c("identity", "nan"),
                          invert = FALSE, n_cores = NULL, progress = FALSE) {
  out_of_bounds <- match.arg(out_of_bounds)
  .elastix_ptr(xf)$xform(
    .xform_xyz(xyz),
    out_of_bounds,
    .invert_flags(invert, length(xf$path), "transform"),
    if (is.null(n_cores)) NULL else as.integer(n_cores),
    isTRUE(progress)
  )
}

#' Invert an Elastix transform on points
#'
#' Elastix itself cannot do this: `transformix` only goes forwards, which is why packages that
#' use Elastix ship two separate registration files per brain pair.
#'
#' Linear steps are inverted exactly. Each B-spline warp has no closed-form inverse and is solved
#' per point by damped Gauss-Newton against the analytic Jacobian.
#'
#' What is guaranteed is **forward-consistency**: `elastix_xform(xf, elastix_xform_inv(xf, y))`
#' returns `y`, to within `accuracy`. What is *not* guaranteed is that inverting a forward
#' transform returns the point you started from, because a B-spline warp need not be injective --
#' a strongly folded registration maps several points to the same place, and no inverse can
#' recover which one you meant. Points with no preimage at all come back as `NaN`.
#'
#' @inheritParams elastix_xform
#' @param initial_guess An `(N, 3)` matrix of starting points for the solver, in the *source*
#'   space. Rarely needed -- the solver seeds itself -- but it is what breaks the tie where a
#'   target has more than one preimage.
#' @param max_iter Solver budget per point.
#' @param seed_iter Rounds of the fixed-point pre-seed. Zero starts the solver at the target,
#'   which loses points wherever the deformation is large.
#' @param tolerance Step-size convergence threshold.
#' @param accuracy Accept a solution only if its residual is within this of the target, in world
#'   units. Otherwise the row is `NaN`.
#' @param lattice_points Size of the global seed lattice -- the last-resort start for the few
#'   points the cheap seeds fail on. Built once per call and consulted only by points that have
#'   already failed. Set to 0 to disable.
#' @param invert The same per-hop flags as on [elastix_xform()], composed with this whole-chain
#'   inversion: hop `i` runs *forwards* here exactly when `invert[i]` is `TRUE`.
#'
#' @return An `(N, 3)` matrix of coordinates in the source space. Rows with no preimage are `NaN`.
#'
#' @seealso [elastix_xform()]
#'
#' @export
elastix_xform_inv <- function(xf, xyz, out_of_bounds = c("identity", "nan"),
                              initial_guess = NULL, max_iter = 50L, seed_iter = 8L,
                              tolerance = 1e-9, accuracy = 1e-3, lattice_points = 16000L,
                              invert = FALSE, n_cores = NULL, progress = FALSE) {
  out_of_bounds <- match.arg(out_of_bounds)
  xyz <- .xform_xyz(xyz)

  # Ask before calling in. extendr cannot carry a Rust panic's message across to R -- it arrives
  # as the useless "User function panicked: xform_inv" -- so this has to be checked here.
  if (!.elastix_ptr(xf)$invertible()) {
    stop(
      "this transform cannot be inverted: it contains an `Add` step ",
      "(HowToCombineTransforms \"Add\"), which does not decompose into invertible hops"
    )
  }

  if (!is.null(initial_guess)) {
    initial_guess <- .xform_xyz(initial_guess, "initial_guess")
    if (nrow(initial_guess) != nrow(xyz)) {
      stop(sprintf(
        "`initial_guess` must have one point per input point: expected %d, got %d",
        nrow(xyz), nrow(initial_guess)
      ))
    }
  }

  .elastix_ptr(xf)$xform_inv(
    xyz,
    out_of_bounds,
    initial_guess,
    as.integer(max_iter),
    as.integer(seed_iter),
    as.double(tolerance),
    as.double(accuracy),
    as.integer(lattice_points),
    .invert_flags(invert, length(xf$path), "transform"),
    if (is.null(n_cores)) NULL else as.integer(n_cores),
    isTRUE(progress)
  )
}

#' Properties of an Elastix transform
#'
#' @param xf An `elastix_transform`, from [elastix_read()].
#'
#' @return
#' * `elastix_affine`: the `4 x 4` matrix of the first linear step, or `NULL` if there is none.
#' * `elastix_kinds`: one string per transform in the chain, giving its resolved steps
#'   initial-first (e.g. `"linear+bspline"`).
#' * `elastix_grid_size` / `elastix_grid_spacing` / `elastix_grid_origin`: a `(k, 3)` matrix, one
#'   row per B-spline in the chain, or `NULL` if the chain is purely linear.
#'
#' @name elastix_properties
#' @export
elastix_affine <- function(xf) .elastix_ptr(xf)$affine()

#' @rdname elastix_properties
#' @export
elastix_kinds <- function(xf) .elastix_ptr(xf)$kinds()

#' @rdname elastix_properties
#' @export
elastix_grid_size <- function(xf) .elastix_ptr(xf)$grid_size()

#' @rdname elastix_properties
#' @export
elastix_grid_spacing <- function(xf) .elastix_ptr(xf)$grid_spacing()

#' @rdname elastix_properties
#' @export
elastix_grid_origin <- function(xf) .elastix_ptr(xf)$grid_origin()


#' Can an Elastix transform be inverted?
#'
#' Answered **without reading the transform's coefficients**, and so cheaply enough to ask about
#' many files at once.
#'
#' A transform cannot be inverted exactly when some step in its chain combines via `Add`:
#' `T(x) = T_initial(x) + T_this(x) - x` does not decompose into invertible hops. That fact lives
#' in one short key -- but the key sits *after* a coefficient array that runs to tens of megabytes,
#' so answering it honestly used to mean parsing the whole file.
#'
#' This walks the same chain and applies the same validation, skipping only the numbers: roughly
#' **20x** faster than [elastix_read()] across real registrations, and ~200x on the largest.
#'
#' Use it when you must label *many* files up front and cannot afford to parse them -- building a
#' bridging graph, say, where each edge needs to know whether it can be walked backwards.
#'
#' @param path Path to a `TransformParameters.*.txt` file. Its initial-transform chain is
#'   followed, however deep.
#'
#' @return `TRUE` if [elastix_xform_inv()] would work on it.
#'
#'   Errors if the file would not load at all (missing, not an Elastix transform, an unsupported
#'   transform kind, binary parameters, a circular chain) -- so `TRUE` is a promise, not a guess.
#'
#' @seealso [elastix_read()], [elastix_xform_inv()]
#'
#' @examples
#' \dontrun{
#' elastix_probe_invertible("TransformParameters.FixedFANC.txt")
#' }
#'
#' @export
elastix_probe_invertible <- function(path) {
  path <- as.character(path)
  if (length(path) != 1L) stop("`path` must name exactly one transform")
  if (!file.exists(path) || dir.exists(path)) {
    stop(sprintf("no Elastix transform file at '%s'", path))
  }
  probe_invertible_raw(path)
}
